### PR TITLE
feat: introduce iterator for default_ethereum_payload function

### LINKED ISSUE #11889: Make payload_builder function accept BestTxterator directly
### Describe the feature

Currently we create the iterator like this:

https://github.com/paradigmxyz/reth/blob/a4126b3a53f900483fe8946c8067aade68cf5cad/crates/ethereum/payload/src/lib.rs#L155-L158

in:
https://github.com/paradigmxyz/reth/blob/a4126b3a53f900483fe8946c8067aade68cf5cad/crates/ethereum/payload/src/lib.rs#L126-L136

it would be nicer if this would accept any iterator

we can make this easily work by changing the requested type of the pool

https://github.com/paradigmxyz/reth/blob/a4126b3a53f900483fe8946c8067aade68cf5cad/crates/ethereum/payload/src/lib.rs#L135-L135

which should not be `TxPool` but something that returns a new `BestTransactionsIter` given a `BestTransactionsAttributes`, so something like

```
Pool: FnOnce(BestTransactionsAttributes) -> BestTransactionsIter
```

see also:

https://github.com/paradigmxyz/reth/pull/11890

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by edisontim**
Closes #11889

### CODE REVIEW

**Comment by edisontim** [comment]
@mattsse, is this what you had in mind?

File: crates/ethereum/payload/src/lib.rs [hunk 1]
```diff
@@ -124,16 +147,18 @@ where
 /// and configuration, this function creates a transaction payload. Returns
 /// a result indicating success with the payload or an error in case of failure.
 #[inline]
-pub fn default_ethereum_payload<EvmConfig, Pool, Client>(
+pub fn default_ethereum_payload<EvmConfig, Pool, Client, IteratorClosure>(
```
Comment (mattsse) [inline crates/ethereum/payload/src/lib.rs:150]: let's just name this F
```suggestion
pub fn default_ethereum_payload<EvmConfig, Pool, Client, F>(
```

File: crates/ethereum/payload/src/lib.rs [hunk 2]
```diff
@@ -124,16 +147,18 @@ where
...
     evm_config: EvmConfig,
     args: BuildArguments<Pool, Client, EthPayloadBuilderAttributes, EthBuiltPayload>,
     initialized_cfg: CfgEnvWithHandlerCfg,
     initialized_block_env: BlockEnv,
+    build_best_txs_closure: IteratorClosure,
```
Comment (mattsse) [inline crates/ethereum/payload/src/lib.rs:155]:
```suggestion
    best_txs: IteratorClosure,
```

File: crates/ethereum/payload/src/lib.rs [hunk 3]
```diff
@@ -45,6 +46,9 @@ use revm_primitives::calc_excess_blob_gas;
 use std::sync::Arc;
 use tracing::{debug, trace, warn};
 
+type BestTransactionsIter<Transaction> =
+    Box<dyn BestTransactions<Item = Arc<ValidPoolTransaction<Transaction>>>>;
```
Comment (mattsse) [inline crates/ethereum/payload/src/lib.rs:50]: let's make this generic over pool

`type BestTransactionsFor<Pool> = Box<dyn BestTransactions<Item = Arc<ValidPoolTransaction<<Pool as TransactionPool>::Transaction>>>>`

File: crates/ethereum/payload/src/lib.rs [hunk 4]
```diff
@@ -94,27 +98,46 @@ where
         args: BuildArguments<Pool, Client, EthPayloadBuilderAttributes, EthBuiltPayload>,
     ) -> Result<BuildOutcome<EthBuiltPayload>, PayloadBuilderError> {
         let (cfg_env, block_env) = self.cfg_and_block_env(&args.config, &args.config.parent_block);
-        default_ethereum_payload(self.evm_config.clone(), args, cfg_env, block_env)
+        let pool = args.pool.clone();
+
+        let iterator_closure = |attributes| pool.best_transactions_with_attributes(attributes);
```
Comment (mattsse) [inline crates/ethereum/payload/src/lib.rs:103]: we don't need a tmp var for this and can make this the argument directly

File: crates/ethereum/payload/src/lib.rs [hunk 5]
```diff
@@ -94,27 +98,46 @@ where
...
-        default_ethereum_payload(self.evm_config.clone(), args, cfg_env, block_env)?
-            .into_payload()
-            .ok_or_else(|| PayloadBuilderError::MissingPayload)
+
+        default_ethereum_payload(
+            self.evm_config.clone(),
+            args,
+            cfg_env,
+            block_env,
+            iterator_closure,
```
Comment (mattsse) [inline crates/ethereum/payload/src/lib.rs:137]: same here

**Review by mattsse — CHANGES_REQUESTED** [review]
cool, only have style nits

**Review by mattsse — APPROVED** [review]
great

**Pull request merged**